services/src/main/java/com/chatbot/services/pubsubservices/PubSubController.java

```diff
@@ -0,0 +1,125 @@
+import com.chatbot.services.protobuf.TriggerEventNotificationOuterClass.TriggerEventNotification;
+import com.chatbot.services.protobuf.TriggerEventNotificationOuterClass.TriggerEventNotification.ChatClient;
+import com.chatbot.services.protobuf.TriggerEventNotificationOuterClass.TriggerEventNotification.Event;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.protobuf.Struct;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestHeader;
+import org.springframework.web.bind.annotation.RestController;
+
+// Handle messages from cloud sub using a push subscriber
+@RestController
+public class PubSubController {
+
+  private HangoutsAsyncService asyncService;
+  private PubSubAuth pubSubAuth;
+  private static final Logger logger = LoggerFactory.getLogger(PubSubController.class);
+
+  public PubSubController(HangoutsAsyncService asyncServiceToSet, PubSubAuth pubSubAuthToSet) {
+    asyncService = asyncServiceToSet;
+    pubSubAuth = pubSubAuthToSet;
+  }
+
+  @PostMapping("/pubsub")
+  String onRequest(@RequestHeader final Map<String, String> headers,
+      @RequestBody final JsonNode message) throws Exception {
+    pubSubAuth.verifyRequest(headers);
+    final String messageData =
+        new String(Base64.getDecoder().decode(message.at("/message/data").asText()));
+    if (messageData.equals(ChatServiceConstants.TRIGGER_EVENT_MESSAGE)) {
+      final TriggerEventNotification triggerEventNotification =
+          buildNotificationFromMessage(message);
+      try {
+        asyncService.triggerEventHandler(triggerEventNotification);
+      } catch (final IOException e) {
+        logger.error("Error while handling event trigger", e);
+      }
+    } else {
+      throw new IllegalArgumentException("Unknown type of message received by subscriber");
+    }
+    return "";
+  }
+
+  private TriggerEventNotification buildNotificationFromMessage(final JsonNode message)
+      throws IllegalArgumentException {
+    final TriggerEventNotification.Builder triggerEventNotificationBuilder = 
+        TriggerEventNotification.newBuilder();
+    if(message.at("/message/attributes").has("userID")) {
+      triggerEventNotificationBuilder.setUserID(message.at("/message/attributes/userID").asText());
+    } else {
+      throw new IllegalArgumentException("No userID provided in published message");
+    }
+    if(message.at("/message/attributes").has("chatClient")) {
+      final String chatClient = message.at("/message/attributes/chatClient").asText();
+      switch (chatClient) {
+        case "HANGOUTS":
+          triggerEventNotificationBuilder.setChatClient(ChatClient.HANGOUTS);
+          break;
+        case "WHATSAPP":
+          triggerEventNotificationBuilder.setChatClient(ChatClient.WHATSAPP);
```

Comment:
As we do not have a setup for Whatsapp yet, throw exception

Reply:
updated